lib/ops/opsservice/operationgroup.go

```diff
@@ -133,6 +139,70 @@ func (g *operationGroup) canCreateOperation(operation ops.SiteOperation) error {
 	return nil
 }
 
+func (g *operationGroup) canCreateUpgradeOperation(cluster ops.Site, operation ops.SiteOperation, force bool) error {
+	// Upgrade is only allowed for active and healthy clusters.
+	if cluster.State != ops.SiteStateActive {
+		return trace.CompareFailed(
+			`Upgrade operation can only be triggered for active clusters. This cluster is currently %v.
+Use gravity status to see the cluster status and make sure that the cluster is active and healthy before retrying.`, cluster.State)
+	}
+	// Even if the cluster is active, run a few checks on the last upgrade
+	// operation to make sure it was completed/rolled back properly, to
+	// protect against cases when cluster state is force-reset midway.
+	lastUpgrade, err := ops.GetLastUpgradeOperation(cluster.Key(), g.operator)
+	if err != nil && !trace.IsNotFound(err) {
+		return trace.Wrap(err)
+	}
+	if trace.IsNotFound(err) {
+		return nil
+	}
+	err = g.checkLastOperation(*lastUpgrade)
+	if err != nil {
+		if force { // Last operation checks didn't pass but force flag was provided.
+			log.WithError(err).Warn("Force-creating upgrade operation.")
+			return nil
+		}
+		return trace.Wrap(err)
+	}
+	// All last operation checks passed.
+	return nil
+}
+
+func (g *operationGroup) checkLastOperation(operation ops.SiteOperation) error {
```

Comment:
Made this more or less generic on purpose, so we can potentially check for other operations too, not just last upgrade.